FEA Fused sparse-dense support for `PairwiseDistancesReduction`

jjerphan opened the pull request:
#### Reference Issues/PRs

Comes after https://github.com/scikit-learn/scikit-learn/pull/23515.

Relates to https://github.com/scikit-learn/scikit-learn/issues/22587.

#### What does this implement/fix? Explain your changes.

Add `SparseSparseDatasetsPair`, `SparseDenseDatasetsPair`, `DenseSparseDatasetsPair` to bridge distances computations for pairs of fused dense-sparse datasets.

This allows support for dense-sparse, sparse-dense and sparse-sparse datasets' pairs for a variety of estimators while at the same time improving performance for the existing sparse-sparse case.

Note that this does not implement optimisation for the Euclidean and Squared Euclidean distances for reasons explained in https://github.com/scikit-learn/scikit-learn/pull/23585#issuecomment-1247996669.

#### TODO

 - [x] https://github.com/scikit-learn/scikit-learn/pull/23604
 - [x] Add tests for new error messages
 - [x] Add changelog entry for fused sparse-dense support on estimators
 - [x] Perform benchmark vs. `main` via `NearestNeighbors`

jjerphan committed:
MAINT Implement CSR support for all DistanceMetric

jjerphan committed:
Merge branch 'main' into maint/dist-metrics-csr-support

jjerphan committed:
TST Remove useless guard

jjerphan committed:
TST Skip JaccardDistance on 32bit architecture

jjerphan committed:
MAINT Define dtype alias for sparse matrices indices

jjerphan committed:
MAINT Do not shadow dtype names in Tempita templating

jjerphan committed:
fixup! MAINT Define dtype alias for sparse matrices indices

jjerphan committed:
TST Use cdist and pdist appropriately

jjerphan committed:
Fixups

jjerphan committed:
DOC Improve comments

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
MAINT Wrap of indptr values to support sparse-dense

This is kind of an hack for now.

IMO, it would be better to use a flatiter on a view if possible.
See discussions on: https://groups.google.com/g/cython-users/c/MR4xWCvUKHU

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Apply review comments

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
FIX Various corrections

ogrisel committed:
More interesting boolean data for tests

jjerphan committed:
FIX Make Jaccard, Hamming and Hashing robust to explicit zeros

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
FIX Make the other boolean DistanceMetric also robust to explicit zeros

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
TST Remove xfail for Jaccard on 32bit arch.

jjerphan committed:
Cast to np.float64_t where appropriate

jjerphan committed:
Rename methods and correctly format their signatures

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

jjerphan committed:
fixup! TST Remove xfail for Jaccard on 32bit arch.

jjerphan committed:
FEA CSR support for HaversineDistance

jjerphan committed:
Fix typo

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Do not upcast to 64bit yet keep the same precision

jjerphan committed:
Do use the default rtol

ogrisel committed:
Set rtol explicitly in test_distance_metrics_dtype_consistency

jjerphan committed:
Implement the sparse-dense and the dense-sparse case for c-contiguity

Also do test for c-contiguity.

jjerphan committed:
Add validation on X and Y, accepting CSR as inputs

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

jjerphan committed:
Remove left-overs

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Merge branch 'main' into maint/dist-metrics-csr-support

jjerphan committed:
Merge branch 'main' into maint/pdr-sparse-support

jjerphan committed:
Add support for all combinations of {dense,sparse} datasets pairs

jjerphan committed:
Const-qualify X and Y

jjerphan committed:
Merge branch 'main' into maint/pdr-sparse-support

jjerphan committed:
Only pass Y_norm_squared for the dense-dense case

jjerphan committed:
Update comments

thomasjpfan commented on sklearn/cluster/_birch.py:
```diff
@@ -694,7 +694,13 @@ def predict(self, X):
         """Predict data using the ``centroids_`` of subclusters."""
-        kwargs = {"Y_norm_squared": self._subcluster_norms}
+
+        # The extra `Y_norm_squared` argument for the back-end
```
Did the old backend support `Y_norm_squared`?

jjerphan replied:
Yes.

jjerphan replied:
Yes it does, but your comment indicates that this is not clear: I am going to clarify it.

jjerphan committed:
Pop unused keywords arguments

jjerphan committed:
Remove unused import

jjerphan committed:
DOC Add whats_new entry

jjerphan committed:
Merge branch 'main' into maint/pdr-sparse-support

jjerphan committed:
fixup! Pop unused keywords arguments

jjerphan commented:
Rah, renaming branches creates a new branch, delete the previous one and close its associated PRs. Let's restore it and reopen this PR.

ogrisel commented on doc/whats_new/v1.2.rst:
```diff
@@ -49,6 +49,32 @@ Changes impacting all modules
+  - :class:`sklearn.manifold.TSNE`
+  - :func:`sklearn.manifold.trustworthiness`
+  - :class:`sklearn.semi_supervised.LabelPropagation`
+  - :class:`sklearn.semi_supervised.LabelSpreading`
```
Have you checked that all those estimators accept sparse inputs?

jjerphan replied:
Good catch! I still need to verify that.

jjerphan replied:
I removed the unrelated estimator (the ones who do not support sparse data) in the change-log entry via 4c455ea440839eedd6f4b681d0b6eaa636aee4f4.

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx:
```diff
@@ -91,14 +94,31 @@ cdef class DatasetsPair:
+
+    @classmethod
+    def unpack_csr_matrix(cls, X: csr_matrix):
+        """Ensure getting ITYPE instead of int internally used for CSR matrices."""
```
```suggestion
        """Ensure that the CSR matrix is indexed with SPARSE_INDEX_TYPE."""
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx:
```diff
@@ -91,14 +94,31 @@ cdef class DatasetsPair:
+    @classmethod
+    def unpack_csr_matrix(cls, X: csr_matrix):
+        """Ensure getting ITYPE instead of int internally used for CSR matrices."""
+        X_data = np.asarray(X.data, dtype=DTYPE)
```
```suggestion
        # TODO: leave X.data unchanged once float32 is supported.
        X_data = np.asarray(X.data, dtype=DTYPE)
```

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx:
```diff
@@ -157,8 +181,181 @@ cdef class DenseDenseDatasetsPair(DatasetsPair):
+
+        # Since Y vectors are dense, we can use a single arrays
+        # of indices of self.n_features elements instead of
+        # a self.n_Y × self.n_features matrices.
```
```suggestion
        # a self.n_Y × self.n_features matrix.
```

ogrisel reviewed (commented):
Looking forward to seeing the benchmark results but otherwise this looks good to me. Just a few small questions/comments/suggestions:

jjerphan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx:
```diff
@@ -157,8 +181,181 @@ cdef class DenseDenseDatasetsPair(DatasetsPair):
+        self.n_Y = Y.shape[0]
+        self.Y_data = np.ravel(Y)
+
+        # Since Y vectors are dense, we can use a single arrays
```
```suggestion
        # Since Y vectors are dense, we can use a single array
```

jjerphan commented on doc/whats_new/v1.2.rst:
```diff
@@ -49,6 +49,32 @@ Changes impacting all modules
+- |Enhancement| Support for combinations of dense and sparse datasets pairs
+  for all distance metrics has been added on the following estimators:
+
+  - :func:`sklearn.metrics.pairwise_distances_argmin`
+  - :func:`sklearn.metrics.pairwise_distances_argmin_min`
+  - :class:`sklearn.cluster.AffinityPropagation`
+  - :class:`sklearn.cluster.Birch`
+  - :class:`sklearn.cluster.MeanShift`
+  - :class:`sklearn.cluster.OPTICS`
+  - :class:`sklearn.cluster.SpectralClustering`
+  - :func:`sklearn.feature_selection.mutual_info_regression`
+  - :class:`sklearn.neighbors.KNeighborsClassifier`
+  - :class:`sklearn.neighbors.KNeighborsRegressor`
+  - :class:`sklearn.neighbors.RadiusNeighborsClassifier`
+  - :class:`sklearn.neighbors.RadiusNeighborsRegressor`
+  - :class:`sklearn.neighbors.LocalOutlierFactor`
+  - :class:`sklearn.neighbors.NearestNeighbors`
+  - :class:`sklearn.manifold.Isomap`
+  - :class:`sklearn.manifold.LocallyLinearEmbedding`
+  - :class:`sklearn.manifold.TSNE`
+  - :func:`sklearn.manifold.trustworthiness`
+  - :class:`sklearn.semi_supervised.LabelPropagation`
+  - :class:`sklearn.semi_supervised.LabelSpreading`
```
```suggestion
  - :func:`sklearn.metrics.pairwise_distances_argmin`
  - :func:`sklearn.metrics.pairwise_distances_argmin_min`
  - :class:`sklearn.cluster.AffinityPropagation`
  - :class:`sklearn.cluster.Birch`
  - :class:`sklearn.cluster.SpectralClustering`
  - :class:`sklearn.neighbors.KNeighborsClassifier`
  - :class:`sklearn.neighbors.KNeighborsRegressor`
  - :class:`sklearn.neighbors.RadiusNeighborsClassifier`
  - :class:`sklearn.neighbors.RadiusNeighborsRegressor`
  - :class:`sklearn.neighbors.LocalOutlierFactor`
  - :class:`sklearn.neighbors.NearestNeighbors`
  - :class:`sklearn.manifold.Isomap`
  - :class:`sklearn.manifold.TSNE`
  - :func:`sklearn.manifold.trustworthiness`
```

jjerphan committed:
DOC Update comment and changelog

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
DOC Update comment and changelog

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

thomasjpfan commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pxd:
```diff
@@ -18,4 +20,33 @@ cdef class DenseDenseDatasetsPair(DatasetsPair):
+
+cdef class SparseSparseDatasetsPair(DatasetsPair):
+    cdef:
+        const DTYPE_t[:] X_data
+        const SPARSE_INDEX_TYPE_t[:] X_indices
+        const SPARSE_INDEX_TYPE_t[:] X_indptr
+
+        const DTYPE_t[:] Y_data
+        const SPARSE_INDEX_TYPE_t[:] Y_indices
+        const SPARSE_INDEX_TYPE_t[:] Y_indptr
```
Is the plan to Tempita all of these? Specifically, `DTYPE_t` to become `float32` and `float64` and `SPARSE_INDEX_TYPE_t` to become `int32` and `int64`.

If so, does this extend to all combinations? For example, sparse data with `float32` data and `int64` indices paired with a sparse data with `float64` data and `int32` indices. 🤯

jjerphan replied:
Potentially yes, especially for `DTYPE_t` which will remain fixed datasetpair-wise.

I don't see how to easily handle various the case on `SPARSE_INDEX_TYPE_t` yet/using Cython, and I fear the complex logic and manual templating…


Other languages would have allowed to ease some logic, but going done this road come after broader discussions IMO.

jjerphan commented:
The benchmarks shows that the results are significantly slower in this PR, especially on small datasets.  E.g. comparing using `metric=manhattan` to use the efficient 

<details>
<summary>128 threads</summary>

```
· Creating environments
· Discovering benchmarks
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scikit-learn commit 780d7bbf <maint/pdr-sparse-support> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors                 ok
[ 50.00%] ··· ========= ============ ============= ==============
              --                   n_test / n_features           
              --------- -----------------------------------------
               n_train   1000 / 100   10000 / 100   100000 / 100 
              ========= ============ ============= ==============
                 1000     410±20ms     3.76±0.1s     36.3±0.2s   
                10000     391±10ms     3.60±0.03s    35.0±0.08s  
                100000   1.18±0.07s    11.7±0.02s     1.97±0m    
              ========= ============ ============= ==============

[ 50.00%] · For scikit-learn commit a5d50cf3 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· ...ancesArgKminBenchmark.time_nearest_neighbors         1/9 failed
[100.00%] ··· ========= ============ ============= ==============
              --                   n_test / n_features           
              --------- -----------------------------------------
               n_train   1000 / 100   10000 / 100   100000 / 100 
              ========= ============ ============= ==============
                 1000    34.7±0.3ms     157±1ms      1.37±0.01s  
                10000     160±2ms      1.37±0.02s    13.8±0.03s  
                100000   1.50±0.03s    15.9±0.4s       failed    
              ========= ============ ============= ==============

[100.00%] ···· For parameters: 100000, 100000, 100
               
               
               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [a5d50cf3]       [780d7bbf]
     <main>           <maint/pdr-sparse-support>
+      1.37±0.01s        36.3±0.2s    26.39  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 100000, 100)
+         157±1ms        3.76±0.1s    23.91  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 10000, 100)
+      34.7±0.3ms         410±20ms    11.82  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(1000, 1000, 100)
+      1.37±0.02s       3.60±0.03s     2.63  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 10000, 100)
+      13.8±0.03s       35.0±0.08s     2.54  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 100000, 100)
+         160±2ms         391±10ms     2.45  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(10000, 1000, 100)
-      1.50±0.03s       1.18±0.07s     0.79  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(100000, 1000, 100)
-       15.9±0.4s       11.7±0.02s     0.74  pairwise_argkmin_estimator.PairwiseDistancesArgKminBenchmark.time_nearest_neighbors(100000, 10000, 100)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```


</details>

IMO, the slowdown can be explained mainly by one thing: IIRC when using `perf-top(1)`, most of the time is spent in distance computations, mainly in the `idiv` instruction (on x86), something which is not observed for `main` in which uses `_sparse_manhattan`.

https://github.com/scikit-learn/scikit-learn/blob/4b252fb126e2c81b1ff1d219905d0e46016f014c/sklearn/metrics/_pairwise_fast.pyx#L38

Apart from reduced indirections in `main@_sparse_manhattan`, the logic in `ManhattanDistance.dist_csr` is identical up to the trick to support sparse and dense dataset:

https://github.com/scikit-learn/scikit-learn/blob/b157ac718754a235e443f3f5671fda7d2cd1dbba/sklearn/metrics/_dist_metrics.pyx.tp#L842-L846

introduced to have an array of `n_features` elements solely for `indices`:

https://github.com/scikit-learn/scikit-learn/blob/63c6fe36e72a4e1aa9490201e856eb8361633b06/sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx#L277-L282

This trick uses modulus operations (hence the call to `idiv`) and might cause the main overhead (`idiv` is generally a costly instruction).

---

I see two possible solutions allowing removing the modulus operations:
 - Do not use an array of `n_features` elements for `indices` but one of `n_features × n_samples` elements. This solution is trivial and safe, but costly
 - Keep the array of `n_features` elements for `indices` but change the API of `DistanceMetric.{dist_csr,rdist_csr}` to take pointers and perform shifting in address for `{x1,x2}_indices` when calling `DistanceMetric.{dist_csr,rdist_csr} using an offset defined based on `{x1,x2}_{start,end}`. This solution is rather hacky, but should work and should come with minimum overhead.

jjerphan committed:
MAINT Test second alternative for sparse-dense support

The modulo trick is costly because it uses integer division
under the hood which are up to 80 times slower.

An alternative is to pass the address of the `indices` array and shift
this address at the callers level with the information that we have
there so that the dereferencing works properly.

This commit replaces the modulo trick by this alternative and adapt
documentation and comments in this regard.

See:
https://github.com/scikit-learn/scikit-learn/pull/23585#issuecomment-1190672626

jjerphan commented:
I've opened https://github.com/jjerphan/scikit-learn/pull/15/files to test this second solution, namely:

> Keep the array of n_features elements for indices but change the signature of `DistanceMetric.{dist_csr,rdist_csr}` to take pointers and perform shifting in address for `{x1,x2}_indices` when calling `DistanceMetric.{dist_csr,rdist_csr}` using an offset defined based on `{x1,x2}_{start,end}`. This solution is rather hacky, but should work and should come with minimum overhead.

jjerphan commented:
Before proceeding with further benchmarks, I would wait for https://github.com/scikit-learn/scikit-learn/pull/23865 to be merged, as it will come with adaptation for this PR.

jjerphan committed:
Merge branch 'main' into maint/pdr-sparse-support

jjerphan committed:
Merge branch 'maint/pdr-sparse-support' into alt/feat/pdr-sparse-support

jjerphan commented:
The alternative proposed in https://github.com/jjerphan/scikit-learn/pull/15/ performs better, see https://github.com/jjerphan/scikit-learn/pull/15/#issuecomment-1209149658.

I would:
 - merge https://github.com/scikit-learn/scikit-learn/pull/23865
 - merge https://github.com/scikit-learn/scikit-learn/pull/24043
 - merge this https://github.com/jjerphan/scikit-learn/pull/15/ in this branch PR
 - merge `main` in this PR branch
 - rerun benchmark on this PR branch against `main`

What do you think?

jjerphan committed:
DOC Improve comments and code self-documentation

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
Merge branch 'main' into maint/pdr-sparse-support

This update the branch after the merge of #23865.

jjerphan committed:
Merge branch 'maint/pdr-sparse-support' into alt/feat/pdr-sparse-support

jjerphan commented on sklearn/metrics/tests/test_pairwise_distances_reduction.py:
```diff
@@ -774,6 +778,92 @@ def test_n_threads_agnosticism(
+@pytest.mark.parametrize(
+    "PairwiseDistancesReduction",
+    [PairwiseDistancesArgKmin, PairwiseDistancesRadiusNeighborhood],
+)
+def test_format_agnosticism(
+    global_random_seed,
+    PairwiseDistancesReduction,
+    n_samples,
+    chunk_size,
+    n_features=100,
+    dtype=np.float64,
+):
```
```suggestion
)
@pytest.mark.parametrize("dtype", [np.float64, np.float32])
def test_format_agnosticism(
    global_random_seed,
    PairwiseDistancesReduction,
    n_samples,
    chunk_size,
    n_features=100,
    dtype,
):
```

jjerphan committed:
TST `dtype`-parametrize `test_format_agnosticism`

jjerphan committed:
fixup! TST `dtype`-parametrize `test_format_agnosticism`

thomasjpfan committed:
MNT Pushing data up instead of indices

thomasjpfan committed:
DOC Improve comment

thomasjpfan committed:
REV Revert back to memoryviews for indices

thomasjpfan committed:
DOC Spelling

jjerphan committed:
Merge pull request #16 from thomasjpfan/alt/feat/pdr-sparse-support

MNT Pushing data up instead of indices

jjerphan committed:
Merge pull request #15 from jjerphan/alt/feat/pdr-sparse-support

MAINT Division-less alternative for fused sparse-dense support

ogrisel commented:
The merge of jjerphan/scikit-learn#15 broke the tests. They were not run on the developer forks unfortunately.

jjerphan committed:
TST Suggest logic adaptation for _pairwise_{dense_sparse,sparse_dense}

This fixes those two methods used for testing which were using
the previous trick.

This allocate potentially large `indices` arrays.

This also add missing kwargs for some others tests.

jjerphan commented:
I've extracted part of https://github.com/jjerphan/scikit-learn/pull/15 in https://github.com/scikit-learn/scikit-learn/pull/24223 to solve the problems appearing on some configuration there and make the diff of this PR smaller.

jjerphan commented:
Benchmarks results against `main` after merging #24223 in this branch using using https://github.com/jjerphan/scikit-learn/commit/2106e02c via `benchmarks/maint/pdr-sparse-support`.

Note that:
 - `'manhattan'` was chosen to better compare performances w.r.t observations covered and remarks made in https://github.com/scikit-learn/scikit-learn/pull/23585#issuecomment-1190672626
 - some jobs failed on `main` due to ASV default 500s timeout and hence are not reported at the end. They complete for this PR branch we get in those case around (and possible more than) ×50 speed-ups. 

```
       before           after         ratio
     [68c5ceb5]       [2106e02c]
     <main>           <benchmarks/maint/pdr-sparse-support>
-         274±4ms          121±7ms     0.44  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 1000, 100, 'manhattan', 'auto')
-      3.92±0.01s         227±30ms     0.06  nn.NearestNeighborsBenchmark.time_kneighbors(10000, 1000, 100, 'manhattan', 'auto')
-      3.82±0.02s         143±30ms     0.04  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 10000, 100, 'manhattan', 'auto')
-       39.0±0.2s       1.12±0.07s     0.03  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 100000, 100, 'manhattan', 'auto')
-      43.9±0.09s       1.14±0.01s     0.03  nn.NearestNeighborsBenchmark.time_kneighbors(100000, 1000, 100, 'manhattan', 'auto')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

<details>
<summary>Full ASV trace</summary>

```
 → asv continuous -b NearestNeighborsBenchmark -e upstream/main benchmarks/maint/pdr-sparse-support                                                                                                                                                                        
· Creating environments                                                                                                                                                                                                                                                         
· Discovering benchmarks                                                                                                                                                                                                                                                        
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl..                                                                                                                                                                                                      
·· Installing 2106e02c <benchmarks/maint/pdr-sparse-support> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl.                                                                                                                                                     
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)                                                                                                                                                                                                        
[  0.00%] · For scikit-learn commit 2106e02c <benchmarks/maint/pdr-sparse-support> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· nn.NearestNeighborsBenchmark.time_kneighbors                                                                                                                                  ok
[ 50.00%] ··· ========= =============================== ================================ =================================
              --                                     n_test / n_features / metric / strategy
              --------- --------------------------------------------------------------------------------------------------
               n_train   1000 / 100 / manhattan / auto   10000 / 100 / manhattan / auto   100000 / 100 / manhattan / auto
              ========= =============================== ================================ =================================
                 1000               121±7ms                         143±30ms                         1.12±0.07s
                10000               227±30ms                       1.87±0.09s                        10.0±0.03s
                100000             1.14±0.01s                      11.2±0.03s                         1.86±0m
              ========= =============================== ================================ =================================
.
[ 50.00%] · For scikit-learn commit 68c5ceb5 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl...
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· nn.NearestNeighborsBenchmark.time_kneighbors                                                                                                                          3/9 failed
[100.00%] ··· ========= =============================== ================================ =================================
              --                                     n_test / n_features / metric / strategy
              --------- --------------------------------------------------------------------------------------------------
               n_train   1000 / 100 / manhattan / auto   10000 / 100 / manhattan / auto   100000 / 100 / manhattan / auto
              ========= =============================== ================================ =================================
                 1000               274±4ms                        3.82±0.02s                        39.0±0.2s
                10000              3.92±0.01s                      39.7±0.04s                          failed
                100000             43.9±0.09s                        failed                            failed
              ========= =============================== ================================ =================================

[100.00%] ···· For parameters: 10000, 100000, 100, 'manhattan', 'auto'


               asv: benchmark timed out (timeout 500s)

               For parameters: 100000, 10000, 100, 'manhattan', 'auto'


               asv: benchmark timed out (timeout 500s)

               For parameters: 100000, 100000, 100, 'manhattan', 'auto'


               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [68c5ceb5]       [2106e02c]
     <main>           <benchmarks/maint/pdr-sparse-support>
-         274±4ms          121±7ms     0.44  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 1000, 100, 'manhattan', 'auto')
-      3.92±0.01s         227±30ms     0.06  nn.NearestNeighborsBenchmark.time_kneighbors(10000, 1000, 100, 'manhattan', 'auto')
-      3.82±0.02s         143±30ms     0.04  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 10000, 100, 'manhattan', 'auto')
-       39.0±0.2s       1.12±0.07s     0.03  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 100000, 100, 'manhattan', 'auto')
-      43.9±0.09s       1.14±0.01s     0.03  nn.NearestNeighborsBenchmark.time_kneighbors(100000, 1000, 100, 'manhattan', 'auto')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>

ogrisel commented:
Very nice. Could you please push the merge of #24223 to see the big picture and be able to concurrently review/test this PR?

jjerphan commented:
> Very nice. Could you please push the merge of https://github.com/scikit-learn/scikit-learn/pull/24223 to see the big picture and be able to concurrently review/test this PR?

Shouldn't we wait for @thomasjpfan's review first?

jjerphan commented:
Additional benchmarks results with `metric="euclidean"` and `np.float32` data via https://github.com/jjerphan/scikit-learn/commit/5bcac125 on a machine with 32 physical cores using 32 threads. We get up to at least ×100 speed-ups.

```

       before           after         ratio
     [68c5ceb5]       [5bcac125]
     <main>           <benchmarks/maint/pdr-sparse-support>
-       212±0.2ms         51.6±1ms     0.24  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 1000, 100, 'manhattan', 'auto', <class 'numpy.float32'>)
-       210±0.4ms         50.9±1ms     0.24  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 1000, 100, 'euclidean', 'auto', <class 'numpy.float64'>)
-       210±0.4ms         50.0±1ms     0.24  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 1000, 100, 'manhattan', 'auto', <class 'numpy.float64'>)
-       212±0.2ms         49.7±1ms     0.23  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 1000, 100, 'euclidean', 'auto', <class 'numpy.float32'>)
-         2.11±0s          102±4ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(10000, 1000, 100, 'euclidean', 'auto', <class 'numpy.float64'>)
-         2.11±0s         102±20ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(10000, 1000, 100, 'manhattan', 'auto', <class 'numpy.float32'>)
-         2.11±0s         100±20ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 10000, 100, 'euclidean', 'auto', <class 'numpy.float32'>)
-         2.11±0s         99.1±1ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(10000, 1000, 100, 'euclidean', 'auto', <class 'numpy.float32'>)
-         2.11±0s         98.9±2ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(10000, 1000, 100, 'manhattan', 'auto', <class 'numpy.float64'>)
-         2.12±0s         98.3±1ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 10000, 100, 'euclidean', 'auto', <class 'numpy.float64'>)
-         2.12±0s         97.7±1ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 10000, 100, 'manhattan', 'auto', <class 'numpy.float64'>)
-         2.11±0s         97.2±1ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 10000, 100, 'manhattan', 'auto', <class 'numpy.float32'>)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

<details>
<summary>Full ASV trace</summary>

```
 → asv continuous -b NearestNeighborsBenchmark -e upstream/main benchmarks/maint/pdr-sparse-support
· Creating environments
· Discovering benchmarks
·· Uninstalling from conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl..
·· Building 5bcac125 <benchmarks/maint/pdr-sparse-support> for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl..................................
·· Installing 5bcac125 <benchmarks/maint/pdr-sparse-support> into conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl..
· Running 2 total benchmarks (2 commits * 1 environments * 1 benchmarks)
[  0.00%] · For scikit-learn commit 5bcac125 <benchmarks/maint/pdr-sparse-support> (round 1/1):
[  0.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[ 50.00%] ··· nn.NearestNeighborsBenchmark.time_kneighbors                                                                                                                                                                                                  ok
[ 50.00%] ··· ========= ======== ======================================== ======================================== ======================================== ========================================
              --                                                                                n_features / metric / strategy / dtype
              ------------------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------
               n_train   n_test   100 / manhattan / auto / numpy.float32   100 / manhattan / auto / numpy.float64   100 / euclidean / auto / numpy.float32   100 / euclidean / auto / numpy.float64
              ========= ======== ======================================== ======================================== ======================================== ========================================
                 1000     1000                   51.6±1ms                                 50.0±1ms                                 49.7±1ms                                 50.9±1ms
                 1000    10000                   97.2±1ms                                 97.7±1ms                                 100±20ms                                 98.3±1ms
                 1000    100000                  602±4ms                                  614±3ms                                  595±2ms                                 617±100ms
                10000     1000                   102±20ms                                 98.9±2ms                                 99.1±1ms                                 102±4ms
                10000    10000                   909±7ms                                  1.01±0s                                  909±6ms                                  1.01±0s
                10000    100000                 6.08±0.01s                               6.37±0.3s                                6.10±0.06s                               6.27±0.2s
                100000    1000                   594±2ms                                  605±1ms                                 1.00±0.05s                               986±300ms
                100000   10000                  6.20±0.2s                                 7.82±2s                                 6.05±0.4s                                6.08±0.01s
                100000   100000                  1.01±0m                                  1.02±0m                                  1.01±0m                                  1.01±0m
              ========= ======== ======================================== ======================================== ======================================== ========================================

[ 50.00%] · For scikit-learn commit 68c5ceb5 <main> (round 1/1):
[ 50.00%] ·· Building for conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl....
[ 50.00%] ·· Benchmarking conda-py3.9-cython-joblib-numpy-scipy-threadpoolctl
[100.00%] ··· nn.NearestNeighborsBenchmark.time_kneighbors                                                                                                                                                                                        12/36 failed
[100.00%] ··· ========= ======== ======================================== ======================================== ======================================== ========================================
              --                                                                                n_features / metric / strategy / dtype
              ------------------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------
               n_train   n_test   100 / manhattan / auto / numpy.float32   100 / manhattan / auto / numpy.float64   100 / euclidean / auto / numpy.float32   100 / euclidean / auto / numpy.float64
              ========= ======== ======================================== ======================================== ======================================== ========================================
                 1000     1000                  212±0.2ms                                210±0.4ms                                212±0.2ms                                210±0.4ms
                 1000    10000                   2.11±0s                                  2.12±0s                                  2.11±0s                                  2.12±0s
                 1000    100000                 21.2±0.03s                               21.2±0.01s                               21.2±0.03s                                21.2±0s
                10000     1000                   2.11±0s                                  2.11±0s                                  2.11±0s                                  2.11±0s
                10000    10000                   21.4±0s                                  20.9±0s                                 21.4±0.01s                                20.9±0s
                10000    100000                   failed                                   failed                                   failed                                   failed
                100000    1000                  21.1±0.01s                               24.3±0.01s                                21.1±0s                                 24.3±0.01s
                100000   10000                    failed                                   failed                                   failed                                   failed
                100000   100000                   failed                                   failed                                   failed                                   failed
              ========= ======== ======================================== ======================================== ======================================== ========================================

[100.00%] ···· For parameters: 10000, 100000, 100, 'manhattan', 'auto', <class 'numpy.float32'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 10000, 100000, 100, 'manhattan', 'auto', <class 'numpy.float64'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 10000, 100000, 100, 'euclidean', 'auto', <class 'numpy.float32'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 10000, 100000, 100, 'euclidean', 'auto', <class 'numpy.float64'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 100000, 10000, 100, 'manhattan', 'auto', <class 'numpy.float32'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 100000, 10000, 100, 'manhattan', 'auto', <class 'numpy.float64'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 100000, 10000, 100, 'euclidean', 'auto', <class 'numpy.float32'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 100000, 10000, 100, 'euclidean', 'auto', <class 'numpy.float64'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 100000, 100000, 100, 'manhattan', 'auto', <class 'numpy.float32'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 100000, 100000, 100, 'manhattan', 'auto', <class 'numpy.float64'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 100000, 100000, 100, 'euclidean', 'auto', <class 'numpy.float32'>


               asv: benchmark timed out (timeout 500s)

               For parameters: 100000, 100000, 100, 'euclidean', 'auto', <class 'numpy.float64'>


               asv: benchmark timed out (timeout 500s)

       before           after         ratio
     [68c5ceb5]       [5bcac125]
     <main>           <benchmarks/maint/pdr-sparse-support>
-       212±0.2ms         51.6±1ms     0.24  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 1000, 100, 'manhattan', 'auto', <class 'numpy.float32'>)
-       210±0.4ms         50.9±1ms     0.24  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 1000, 100, 'euclidean', 'auto', <class 'numpy.float64'>)
-       210±0.4ms         50.0±1ms     0.24  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 1000, 100, 'manhattan', 'auto', <class 'numpy.float64'>)
-       212±0.2ms         49.7±1ms     0.23  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 1000, 100, 'euclidean', 'auto', <class 'numpy.float32'>)
-         2.11±0s          102±4ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(10000, 1000, 100, 'euclidean', 'auto', <class 'numpy.float64'>)
-         2.11±0s         102±20ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(10000, 1000, 100, 'manhattan', 'auto', <class 'numpy.float32'>)
-         2.11±0s         100±20ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 10000, 100, 'euclidean', 'auto', <class 'numpy.float32'>)
-         2.11±0s         99.1±1ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(10000, 1000, 100, 'euclidean', 'auto', <class 'numpy.float32'>)
-         2.11±0s         98.9±2ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(10000, 1000, 100, 'manhattan', 'auto', <class 'numpy.float64'>)
-         2.12±0s         98.3±1ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 10000, 100, 'euclidean', 'auto', <class 'numpy.float64'>)
-         2.12±0s         97.7±1ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 10000, 100, 'manhattan', 'auto', <class 'numpy.float64'>)
-         2.11±0s         97.2±1ms     0.05  nn.NearestNeighborsBenchmark.time_kneighbors(1000, 10000, 100, 'manhattan', 'auto', <class 'numpy.float32'>)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

</details>

jjerphan committed:
Merge branch 'main' into maint/pdr-sparse-support

jjerphan committed:
DOC Add co-authors in `whats_new` entry

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>
Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
Do not support CSR matrices without non-zero elements

jjerphan committed:
Merge branch 'main' into maint/pdr-sparse-support

jjerphan committed:
fixup! Merge branch 'main' into maint/pdr-sparse-support

jjerphan committed:
MAINT Do not pop Y_norm_squared when unused

jjerphan commented:
https://github.com/scikit-learn/scikit-learn/pull/24272 has been opened to resolve test failures observed in this PR.

ogrisel commented on sklearn/metrics/_pairwise_distances_reduction/_argkmin.pyx.tp:
```diff
@@ -98,14 +98,8 @@ cdef class ArgKmin{{name_suffix}}(BaseDistanceReducer{{name_suffix}}):
-            # supported for the Euclidean variant.
-            if metric_kwargs is not None:
-                metric_kwargs.pop("Y_norm_squared", None)
-
```
Maybe we should instead change the upstream estimator objects to avoid passing precomputed `Y_norm_squared` when the train or test data is sparse.

If it's too complicated, then I am fine with keeping the `pop` trick.

jjerphan replied:
Resolving this conversation as we agreed on a solution in #24272.

jjerphan committed:
Merge branch 'main' into maint/pdr-sparse-support

jjerphan committed:
Explicitly do not support CSR matrices with int64 indices and indptr

See: https://github.com/scikit-learn/scikit-learn/issues/23653

jjerphan committed:
DOC Update and improve comment for the alternative CSR representation

jjerphan committed:
fixup! DOC Update and improve comment for the alternative CSR representation

jjerphan committed:
CI Retrigger CI due to faulty runs

jjerphan committed:
Merge branch 'main' into maint/pdr-sparse-support

jjerphan committed:
DOC Update whats_new entry

jjerphan committed:
Test and document Isomap on sparse data

jjerphan committed:
Test and document TSNE on sparse data

jjerphan committed:
Test and document pairwise_distances_argmin on sparse data

jjerphan committed:
Test and document LocalOutlierFactor on sparse data

jjerphan committed:
DOC Add support for sparse data for NearestNeighbors, KNeighbors*, RadiusNeighbors* methods

jjerphan commented on doc/whats_new/v1.2.rst:
```diff
@@ -305,7 +328,7 @@ Changelog
    the :func:`metrics.label_ranking_average_precision_score` metric.
    :pr:`23442` by :user:`Sean Atukorala <ShehanAT>`
-  
+
```
```suggestion
  
```

jjerphan committed:
DOC Remove formatting change

jjerphan commented:
I just have added minimal changes to tests and documentation for user API.

I think we can tests combinations of sparse and dense datasets more thoroughly and systematically. This necessitates refactoring the tests, which I think we might prefer to in another PR.

I think this PR is mergeable.

ogrisel reviewed (commented):
A few comments below but otherwise the diff looks good to me.

However when experimenting with this PR locally I found a quite significant performance regression in `LocalOutlierFactor` and `Birch` on sparse float32 data (I am not yet systematically tested the others), typically around 1.2x to 1.5x slower.

For instance for Birch:

- on main:

```python
In [1]: from sklearn.cluster import Birch
   ...: from scipy import sparse as sp
   ...: import numpy as np
   ...: 
   ...: X = sp.random(5000, 1000, density=0.01, format="csr", random_state=0).astype(np.float32)
   ...: %time Birch().fit(X)
CPU times: user 41.6 s, sys: 1.15 s, total: 42.7 s
Wall time: 10.9 s
```

- on this branch:

```python
In [1]: from sklearn.cluster import Birch
   ...: from scipy import sparse as sp
   ...: import numpy as np
   ...: 
   ...: X = sp.random(5000, 1000, density=0.01, format="csr", random_state=0).astype(np.float32)
   ...: %time Birch().fit(X)
CPU times: user 1min 24s, sys: 619 ms, total: 1min 25s
Wall time: 12.6 s
```

and for LOF:

- on `main`:


```python
In [1]: from sklearn.neighbors import LocalOutlierFactor
   ...: from scipy import sparse as sp
   ...: import numpy as np
   ...: 
   ...: X = sp.random(50000, 1000, density=0.01, format="csr", random_state=0).astype(np.float32)
   ...: %time lof = LocalOutlierFactor().fit(X)
   ...: lof.negative_outlier_factor_
CPU times: user 25.3 s, sys: 3.21 s, total: 28.5 s
Wall time: 28.5 s
Out[1]: 
array([-2.25244  , -3.163138 , -3.0220103, ..., -3.5005074, -3.367448 ,
       -2.2913156], dtype=float32)
```

- on this branch:

```python
In [1]: from sklearn.neighbors import LocalOutlierFactor
   ...: from scipy import sparse as sp
   ...: import numpy as np
   ...: 
   ...: X = sp.random(50000, 1000, density=0.01, format="csr", random_state=0).astype(np.float32)
   ...: %time lof = LocalOutlierFactor().fit(X)
   ...: lof.negative_outlier_factor_
CPU times: user 4min 48s, sys: 698 ms, total: 4min 49s
Wall time: 37.5 s
Out[1]: 
array([-2.25244047, -3.16313819, -3.02201071, ..., -3.50050721,
       -3.36744845, -2.29131578])
```

Also not that the `.negative_outlier_factor_` fitted attribute used to be `float32` on `main` while it is upcasted to `float64` in this branch.

jjerphan commented:
The previous back-end is sometimes more performant because it manages to use the same decomposition for chunks of the Squared Euclidean distance matrix used GRMM in the dense case, namely:

$$
\mathbf D^{\odot 2}_d \left(\mathbf{X}_{c}^{(l)}, \mathbf{Y}_{c}^{(k)}\right) \overset{\text{euclidean case}}{=} \left[\left\Vert \left({\mathbf{X}_c^{(l)}}\right)_{i\cdot}\right\Vert^2_2 \right]_{(i,j)\in [c]\times[c]} + \left[\left\Vert \left({\mathbf{Y}_c^{(k)}}\right)_{j\cdot }\right\Vert^2_2 \right]_{(i, j)\in [c]\times[c]} - 2 \mathbf{X}_c^{(l)} {\mathbf{Y}_c^{(k)}}^\top
$$

where the two first terms (likely) are computed once and are reused and the last is computed via `safe_sparse_dot`, e.g. `euclidean_distances` for the sparse float32 case:

https://github.com/scikit-learn/scikit-learn/blob/97231159ee7a21627e6cb7dc406d611850c465ce/sklearn/metrics/pairwise.py#L568-L570

In this case, `safe_sparse_dot` calls `scipy._csr_array.__matmul__` which ultimately routes to [the `csr_matvecs` C++ routine](https://github.com/scipy/scipy/blame/8e228ffe6330978ac8a2aaa3d8ed67a342040646/scipy/sparse/sparsetools/csr.h#L1139-L1175).

I think we might be able to use the `Euclidean` specialisation for the fused {sparse, dense}² by extending the `GEMMTermComputer` case to call relevant SciPy's C++ routines (in this case `GEMMTermComputer` will be renamed because GEMM won't be called since sparse data will be supported used).

I would prefer exploring this in another PR and in the meantime just deactivate the new implementations proposed in this PR for user-facing APIs that are subjects to regressions (using a `sklearn.config_context`-context manager).

<details>
<summary>`py-spy` profile</summary>

```python
# script.py
from sklearn.cluster import Birch
from scipy import sparse as sp
import numpy as np

X = sp.random(5000, 1000, density=0.01, format="csr", random_state=0).astype(np.float32)
Birch().fit(X)
```

```bash
py-spy record --native -o py-spy-`git rev-parse HEAD`.svg -f speedscope -- python ./script.py
```

On main: ![py-spy-profile-851d02f2ae0dbf258f7cd2f1214ba47c79d1737a](https://user-images.githubusercontent.com/13029839/190395741-935425fd-05f4-4a16-bd40-797bffed1c32.svg)


On this branch: ![py-spy-profile-1eb5b2c3ce31f96eff0922adbdd236c5f0a5c879](https://user-images.githubusercontent.com/13029839/190395757-46f22c8d-fb7d-47de-9ff1-d50bdf33e264.svg)


To browse on https://www.speedscope.app/


</details>

jjerphan committed:
TST Do not test on full cartesian product

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
fixup! TST Do not test on full cartesian product

jjerphan committed:
TST Add TODO for consistency checks on results for sparse and dense data

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
MAINT Mark PairwiseDistancesReductions as unusable for some config.

jjerphan committed:
fixup! MAINT Mark PairwiseDistancesReductions as unusable for some config.

ogrisel commented:
> What do you think?

Thanks for your investigations. I agree with your plan.

Before merging this, what do you think of https://github.com/scikit-learn/scikit-learn/pull/23585#discussion_r968419571 ?

jjerphan commented:
> Before merging this, what do you think of https://github.com/scikit-learn/scikit-learn/pull/23585#discussion_r968419571 ?

Oh, I already have accepted this sensible suggestion in fcf15b65fc51461d43c3bb14d78f2236f06d44fb. :slightly_smiling_face:

ogrisel reviewed (approved):
LGTM. I did some manual tests quickly with the updated PR and did not observe any regression.

Micky774 commented on sklearn/metrics/_pairwise_distances_reduction/_datasets_pair.pyx.tp:
```diff
@@ -172,9 +195,211 @@ cdef class DenseDenseDatasetsPair{{name_suffix}}(DatasetsPair{{name_suffix}}):
+
+    @final
+    cdef ITYPE_t n_samples_Y(self) nogil:
+        return self.Y_indptr.shape[0] -1
```
```suggestion
        return self.Y_indptr.shape[0] - 1
```

Micky774 reviewed (approved):
Sorry for taking so long to look at this. Overall looks great! I just left a minor suggestion for simplifying the tests a bit, and a couple of optional nits.

Edit: Should this also remove the comments here:
https://github.com/scikit-learn/scikit-learn/blob/fec55bf54418c85ffc7b4ea742994b3e57964e09/sklearn/metrics/pairwise.py#L691-L692

and here:
https://github.com/scikit-learn/scikit-learn/blob/fec55bf54418c85ffc7b4ea742994b3e57964e09/sklearn/metrics/pairwise.py#L802-L803

jjerphan committed:
TST Improve test_format_agnosticism

…and add a space somewhere for proper formatting.

Co-authored-by: Meekail Zain <Micky774@users.noreply.github.com>

jjerphan committed:
DOC Update comment regarding the use of pairwise_distances_chunked

See: `BaseDistanceReductionDispatcher.valid_metrics`

Co-authored-by: Meekail Zain <Micky774@users.noreply.github.com>

ogrisel commented:
Merged! Thank you @jjerphan and everybody else!

jjerphan commented:
Thank you for the reviews, @ogrisel, @thomasjpfan and @Micky774!
